File: sdks/node/src/domain/service/stateless_determinism.ts
import type { IDeterminism } from '../core/deterministic.js';

class StatelessDeterminism implements IDeterminism {
  constructor(readonly states: Record<string, string>) {}

  get(key: string, origin: () => string): string {
    if (!this.states) throw new Error('NullReferenceException: States dictionary is null');

    const state = this.states[key];
    if (state != null) {
      return state;
    }

    const val = origin();
    this.states[key] = val;
    return val;
  }
}

export { StatelessDeterminism };

File: sdks/dotnet/sulfone-helium/Domain/Service/StatelessDeterminism.cs
using System.Text.Json;
using sulfone_helium.Domain.Core;

namespace sulfone_helium.Domain.Service;

public class StatelessDeterminism(Dictionary<string, string> states) : IDeterminism
{
    public Dictionary<string, string> States { get; } = states;

    public string Get(string key, Func<string> origin)
    {
        if (States == null)
            throw new NullReferenceException("States dictionary is null");

        if (States.TryGetValue(key, out var state))
            return state;

        var val = origin();
        States[key] = val;
        Console.WriteLine(JsonSerializer.Serialize(States));
        return val;
    }
}
